atomix/cluster/src/main/java/io/atomix/raft/roles/PassiveRole.java

```diff
@@ -531,6 +531,7 @@ protected void appendEntries(
 
         final boolean failedToAppend = tryToAppend(future, reader, entry, index, lastEntry);
         if (failedToAppend) {
+          flush(lastLogIndex - 1, request.prevLogIndex());
```

Comment:
The future is already completed before the log is flushed. That means, the follower acknowledges before the entries are flushed. `failAppend()` must be called after this flush. Can you check if it possible to refactor `tryToAppend` and take `failAppend` out of it and move it here?

Reply:
It's completed but not returned so the ack happens after the flush

Reply:
Ya. You are right. The usage for future there is confusing. It is a synchronous execution even though it uses a future :slightly_smiling_face:

Reply:
I agree, it's a bit strange to use a future but return it synchronously. It's a good candidate for a future refactor